src/CoreWCF.Primitives/src/CoreWCF/Security/MessageSecurityOverHttp.cs

```diff
@@ -0,0 +1,437 @@
+        {
+            clientCredentialType = DefaultClientCredentialType;
+            negotiateServiceCredential = DefaultNegotiateServiceCredential;
+            algorithmSuite = SecurityAlgorithmSuite.Default;
+        }
+
+        public MessageCredentialType ClientCredentialType
+        {
+            get { return this.clientCredentialType; }
+            set
+            {
+                if (!MessageCredentialTypeHelper.IsDefined(value))
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new ArgumentOutOfRangeException("value"));
+                }
+                this.clientCredentialType = value;
+            }
+        }
+
+        public bool NegotiateServiceCredential
+        {
+            get { return this.negotiateServiceCredential; }
+            set { this.negotiateServiceCredential = value; }
+        }
+
+        public SecurityAlgorithmSuite AlgorithmSuite
+        {
+            get { return this.algorithmSuite; }
+            set
+            {
+                if (value == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("value");
+                }
+                this.algorithmSuite = value;
+                wasAlgorithmSuiteSet = true;
+            }
+        }
+
+        internal bool WasAlgorithmSuiteSet
+        {
+            get { return this.wasAlgorithmSuiteSet; }
+        }
+
+        protected virtual bool IsSecureConversationEnabled()
+        {
+            return true;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public SecurityBindingElement CreateSecurityBindingElement(bool isSecureTransportMode, bool isReliableSession, MessageSecurityVersion version)
+        {
+            if (isReliableSession && !this.IsSecureConversationEnabled())
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecureConversationRequiredByReliableSession)));
+            }
+
+            SecurityBindingElement result;
+            SecurityBindingElement oneShotSecurity = null;
+
+            bool isKerberosSelected = false;
+            bool emitBspAttributes = true;
+            if (isSecureTransportMode)
+            {
+                switch (this.clientCredentialType)
+                {
+                    case MessageCredentialType.None:
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.ClientCredentialTypeMustBeSpecifiedForMixedMode)));
+                    case MessageCredentialType.UserName:
+                        oneShotSecurity = SecurityBindingElement.CreateUserNameOverTransportBindingElement();
+                        break;
+                    //case MessageCredentialType.Certificate:
+                    //    oneShotSecurity = SecurityBindingElement.CreateCertificateOverTransportBindingElement();
```

Comment:
We should aim to support Certificate too. Supporting two different certificate types ensures that nothing is hard coded for the single use case.

Reply:
ported.. need help with test case